### ISSUE TITLE
Possibly more robust data passing instead of delimiting with nbsp

### ISSUE THREAD

**Issue by jaekyeom** [ibhagwan/fzf-lua#189]
Hi,
I'm more than satisfied by this great plugin. Thank you so much.

Recently, I was working with some files and found that fzf-lua didn't handle them correctly while fzf.vim did.
I checked and it turned out that those files had the nbsp (" ", 00a0) character in their names.
I'm not that familiar with the internals of fzf-lua yet, but do you think there can be another way to achieve the same thing without using some character as a delimiter?

**Comment by jaekyeom** [comment]
Thanks for the suggestion.

That sounds good as a workaround. If it allows specifying a string as a delimiter (e.g., concatenation of two different invisible characters), then I would be able to lower the possibility of the delimiter appearing in random files enough.

**Comment by ibhagwan** [comment]
> Thanks for the suggestion.
> 
> That sounds good as a workaround. If it allows specifying a string as a delimiter (e.g., concatenation of two different invisible characters), then I would be able to lower the possibility of the delimiter appearing in random files enough.

I’ll give it some more thought, two invisible chars might work but it will look bad as you’ll have an extra “space” between the icons and the filename.

**Comment by jaekyeom** [comment]
Or, how about splitting the result only on the first occurrence of the nbsp character?
There may be some cases that the splitting should be done on the first n occurrences (e.g., buffers?), but it will still work, I guess?

**Comment by ibhagwan** [comment]
> Or, how about splitting the result only on the first occurrence of the nbsp character? There may be some cases that the splitting should be done on the first n occurrences (e.g., buffers?), but it will still work, I guess?

It's not only buffers, it's also the spacing between the file icon and the git icon.

I was busy with a pretty big change that was just committed so I can now attend to this issue, why don't we also look into where exactly does it fail perhaps we can apply the fix there rather than changing the internal logic.

I would assume this only affects the previewer and actions for the files that has `0xA0` in their names? Are there any other issues this causes?

I'm assuming the below line is what causes the entire issue, which does a "greedy match" of "anything after the last occurrence of `nbsp`":

https://github.com/ibhagwan/fzf-lua/blob/f947669dec0612644ff58afec28e003c97e1c733/lua/fzf-lua/path.lua#L142

**Comment by ibhagwan** [comment]
> Yes, for the nbsp issue, I think it is the exact cause.
> 
> A different but somewhat similar issue: I think files with ":" in their names cause a similar problem.

Well the assumption is that `:` is the `rg|grep` separator for the context (`file.extension:line:col:text`), obviously anything can be solved but is this really a legit use case? does anyone willingly use `:` within their filenames?

As for the `nbsp` I'm thinking to just add a counter per function of the number of `nbsp`'s and then strip the strip the string up to the nth `nbsp`.

**Comment by jaekyeom** [comment]
> does anyone willingly use : within their filenames?

Maybe not. I haven't seen such cases yet.
(For future reference, I did a quick check on vim's `getqflist` function. It works like scanf and handles some more cases where ":" is in the filenames but not all.)

> As for the nbsp I'm thinking to just add a counter per function of the number of nbsp's and then strip the string up to the nth nbsp.

That sounds like a good way to go!

**Comment by jaekyeom** [comment]
> The problem here is not just : in the filename, that's pretty easy to solve when it's only a filename, the problem is on more complex output similar to what rg matches print, there you have : in both the filename and as a separator, at which point you need to split at :<digit>.

Yes, that's why I looked into `getqflist` to see how it handles such things. It turned out it also fails when filenames contain things like ":10:3".

> it's not just fzf-lua functions, some users also wrote extensions and other cool stuff, requiring them to set it isn't very clean nor intuitive IMO.

A fair point, I was not aware of that.
Thanks for your effort.

**Comment by ibhagwan** [comment]
> > The problem here is not just : in the filename, that's pretty easy to solve when it's only a filename, the problem is on more complex output similar to what rg matches print, there you have : in both the filename and as a separator, at which point you need to split at :.
> 
> Yes, that's why I looked into `getqflist` to see how it handles such things. It turned out it also fails when filenames contain things like ":10:3".

If the quickfix functions also fail with such names I'm feeling alright about not having the extra complexity to support this :-)


> > it's not just fzf-lua functions, some users also wrote extensions and other cool stuff, requiring them to set it isn't very clean nor intuitive IMO.
> 
> A fair point, I was not aware of that. Thanks for your effort.

I'm exploring 2 different options:
- Striping the nbsp sequences only when they appear at the start of the string: this seems pretty good except that the git icons cause this to fail when the file name starts with a git icon indicator followed by `nbsp` (i.e. the file is named `M name.ext`) - I would assume this is extremely rare, but still, if I'm about to solve it why leave a case where it could fail

- Adding an ansi escape sequence to mark the start of the filename, need to find the right one for that since `rg --color=always` also uses ansi escape sequences around the filename for coloring.

**Comment by jaekyeom** [comment]
Good to know, thank you so much.

**Comment by ibhagwan** [comment]
I wonder why `fzf` doesn't like [ZWS](https://en.wikipedia.org/wiki/Zero-width_space), if we can find a way to display this character in `fzf` properly (and not show `<200b>`) this will be the solution.

**Comment by ibhagwan** [comment]
Alright, good news, I was finally able to solve this properly, used [Unicode character 2060, aka "Word Joiner"](http://www.unicode-symbol.com/u/2060.html).

Try the latest commit and lmk if it works well for you?

**Comment by jaekyeom** [comment]
I've just tried it and experienced the same thing. Fzf showed some weird behaviors matching files.. 😢

**Comment by ibhagwan** [comment]
Good to hear!

Closing.
